### DISCUSSION TITLE
Adding support for stm32 sdhc Zephyr driver

### DISCUSSION THREAD

**Discussion by ivanwagner** [zephyrproject-rtos/zephyr#77079] [General] [answered]
Is there any plan on adding support for sdhc on the stm32 platform?

**Comment by ndrs-pst** [comment]
@ivanwagner  l think it's already here
,🤔   https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers%2Fdisk%2Fsdmmc_stm32.c

**Reply by ivanwagner** [reply]
I was thinking about his specifically: https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/sdhc

What's the difference?

**Reply by ndrs-pst** [reply]
That’s due to the difference between the MMC and SD specifications.

In short, here’s the wording from `stm32h7xx_hal_mmc.c` and `stm32h7xx_hal_sd.c`:

- SDMMC driver functions to interface with MMC and eMMC card devices.
- SDMMC driver functions to interface with SD and uSD card devices.

So, if you're using uSD card devices, then just leave the following `CONFIG_SDMMC_STM32_EMMC`

---
https://github.com/zephyrproject-rtos/zephyr/blob/ad0a1c76bd107831c2b95282f3f6b6f1778b129c/drivers/disk/Kconfig.sdmmc#L65-L69

**Reply by ivanwagner** [reply]
So to my understanding if the case would be a generic SDIO to communicate with a device, would SD be it?

**Reply by ndrs-pst** [reply, marked as answer]
In the context of using an SDIO interface with an external chip, such as a Wi-Fi chipset, I observed that Zephyr's codebase includes a driver for Infineon's AIROC Wi-Fi
https://github.com/zephyrproject-rtos/zephyr/blob/ad3e941ad3ed9e3e2f9032b0151558a948e99fb9/drivers/wifi/infineon/airoc_wifi.h#L15-L19

, which seems to be used in combination with, for example, this SDIO driver

https://github.com/zephyrproject-rtos/zephyr/blob/ad3e941ad3ed9e3e2f9032b0151558a948e99fb9/drivers/sdhc/ifx_cat1_sdio.c#L306-L315

This appears to be the closest match to what you are asking for.

**In conclusion, aside from using an SD card, I'm afraid that STM32 is not yet supported.**

FWIW, https://github.com/Infineon/AIROC-Wi-Fi-Bluetooth-STM32/blob/master/Middlewares/Third_Party/pal/targets/TARGET_STM32/Src/stm32_cyhal_sdio.c

**Reply by ivanwagner** [reply]
Hi @ndrs-pst thank you for your answer the situation is now clear to me.

**Reply by ndrs-pst** [reply]
It's my pleasure to help the community! I'm glad I could assist.

Through this investigation, I realized that the correct place for the STM32 SDMMC should be in
the SDHC subsystem and that it should implement the `sdhc_driver_api` instead of `disk_operations`.

Ping @erwango if you have any plan for this 👀

**Reply by erwango** [reply]
Yes, there are plans, but we're waiting a new HAL SDIO API which wouldn't include SDHC operations as it does today.
This should be completed before eoy.

**Reply by facchinm** [reply]
@erwango as you know we are mildly interested in the topic 😇 any updated estimation?

**Reply by erwango** [reply]
Bad news is that we're fully busy with other topics atm and it won't happen this year unfortunately.
Good news is that hal has been delivered and is now available here: https://github.com/zephyrproject-rtos/hal_stm32, so if anyone is interested ...
Otherwise, we're currently reviewing plans for coming months, we'll try to find a slot.

**Reply by erwango** [reply]
@facchinm We're trying to find a better solution. I'll keep you posted.

**Reply by erwango** [reply]
@facchinm, @ndrs-pst, @ivanwagner: @ExaltZephyr has started development of the driver which should be available in coming weeks.
